### ISSUE TITLE
[Bug] t[:, a:b] on a multi-row tile silently returns wrong data — pl.tile.slice encodes a dense MemRef over its own live source

### ISSUE THREAD

**Issue by luohuan19** [hw-native-sys/pypto#2010]
## Background

**`t[:, a:b]` — a column slice of a MULTI-ROW tile — silently returns wrong data.** It compiles
clean, raises nothing, produces no NaN. `ROWS = 1` is correct; every `ROWS >= 2` is corrupted.

This is about as basic as a DSL operation gets, and it is the fourth silent data-corruption defect
found in this same investigation (after #2003, #2007, #2006).

Found in qwen3-14b's fused decode (`models/qwen3/14b/decode_layer.py`), whose RoPE step column-slices
a `[16, 128]` tile into halves. The kernel returns wrong tokens; the in-repo golden cannot see it
because that golden is a second on-device dispatch of the same code path.

Reproduction environment:

| Component | Version |
|---|---|
| pypto | `8ebddcb8` (`main`) + the fixes for #2003 / #2007 / #2006 — reproduces with all three applied |
| simpler | `c94aa9f3` |
| ptoas | `0.48` |
| pto-isa | `83d01313` (the pin) |
| CANN | `9.0.0` |
| NPU | Ascend 910B (`a2a3`) |

### Component

IR / MemRef (`pl.tile.slice`)

## Root cause

pypto lowers `t[:, a:b]` to `pl.tile.slice` -> `MemRef(parent_buf, byte_offset = a * esize,
size = rows * cols_slice * esize)` — a **flat byte offset into the parent with a DENSE size**.
That encoding is only coherent when `rows == 1`.

Generated `.pto` for the qwen3 case (`q_heads` is `[16, 128]` f32, row pitch 512 B):

```mlir
%q_heads @76032  16x128 f32   (row pitch 512 B)
%q_lo    @76032  16x 64 f32   pad=0   (row pitch 256 B)   <-- base == q_heads
%q_hi    @76288  16x 64 f32   pad=0   (row pitch 256 B)   <-- base == q_heads + 256 B
pto.textract ins(%q_heads, 0,  0) outs(%q_lo)
pto.textract ins(%q_heads, 0, 64) outs(%q_hi)
```

`textract` must repack **strided (pitch 128) -> dense (pitch 64)**, and its destination lands
**on top of its own still-live source**:

- `q_lo`'s extract trashes `q_heads`;
- `q_hi`'s extract then reads the already-trashed `q_heads` **and** overwrites `q_lo` — which is read
  again afterwards.
- Only **row 0** survives, because for row 0 the dense destination address happens to equal its
  source address.

This was simulated byte-for-byte on the host from the kernel's own inputs: the simulation predicts
the device's `q_heads` **and** the downstream `q_rot` with **max|diff| = 0.000000 across all 2048
rows**.

**Why `ROWS = 1` is safe** (and why the qwen3 K path is unaffected while Q is destroyed): with one
row, source and destination pitches coincide, so the repack is an identity copy. qwen3's K path
row-slices to 1 row *first* (`k_normed[0:1, :]`) and only then column-slices; its Q path
column-slices the full 16-row tile. A Q/K bisect on the real kernel isolated the bug to Q exactly
along that line.

## Actual behaviour

```
   ROWS   max_abs_err     wrong elems    wrong rows  verdict
  --------------------------------------------------------------------
      1      0.000000       0/128          0/1       PASS
      2      3.602502      64/256          1/2       *** WRONG ***
      4      7.598506     128/512          2/4       *** WRONG ***
      8      7.048662     256/1024         4/8       *** WRONG ***
     16      9.420689     512/2048         8/16      *** WRONG ***
```

Signature at `ROWS = 16` — `out[0, 64:68]` comes back holding **row 1's lower half**:

```
device   out[0, 64:68]        = [-0.5881, -0.3706, -0.3759, -0.3649]
correct  x[0, 64:68] * w[:4]  = [-0.3789,  0.8139,  0.5189, -0.8323]
x[1, 0:4] * w[0:4]            = [-0.5881, -0.3706, -0.3759, -0.3649]   <-- what actually lands
```

Exactly the dense-repack-over-live-source signature.

## Steps to reproduce

Standalone; imports only `pypto.language` + `pypto.runtime`. No pypto-lib, no golden harness. ~100
lines. **No `pl.spmd`, no `pl.system.syncall`, no `pl.split_aiv`, no `pl.create_tensor`** — a column
slice of a multi-row tile is sufficient on its own.

```bash
python repro_colslice.py -d <device_id>     # exits 1 on the bug
```

<details>
<summary><b>repro_colslice.py</b></summary>

```python
"""STANDALONE repro — pypto only (pypto.language + pypto.runtime). No pypto-lib, no golden
harness.

A COLUMN slice of a MULTI-ROW tile is silently miscompiled.

    t   = x[0:ROWS, :]            # [ROWS, 128] f32 tile in UB, row pitch 128 floats
    lo  = t[:, 0:64]              # -> pto.textract into a DENSE [ROWS,64] tile (pitch 64)
    hi  = t[:, 64:128]            #    ALLOCATED AT t's OWN BASE ADDRESS (+0 / +64 floats)

pypto lowers each column slice to a `pto.textract` that repacks strided(pitch 128) ->
dense(pitch 64), and the memory allocator gives the destination the *subview's* address,
i.e. INSIDE the still-live source tile. The repack is therefore performed IN PLACE over its
own input:

  * lo's extract overwrites t (dest row r lands at t+64r, source row r sits at t+128r);
  * hi's extract then reads the ALREADY-TRASHED t, and its dest (t+64 .. t+64+64*ROWS)
    overwrites lo as well.

Only row 0 survives (its destination address equals its source address). For ROWS == 1 the
whole thing degenerates to an identity copy, so a single-row tile is unaffected — which is
why the same construct is harmless elsewhere.

Run:  python repro_colslice.py -d 3
"""
import argparse

import torch

import pypto.language as pl
from pypto.runtime import RunConfig

COLS = 128
HALF = 64

ap = argparse.ArgumentParser()
ap.add_argument("-d", "--device", type=int, default=3)
args = ap.parse_args()

import pypto.pypto_core as _c  # noqa: E402
print(f"[pypto] {_c.__file__}\n")

cfg = RunConfig(platform="a2a3", device_id=args.device, enable_dep_gen=False)


def build(rows):
    @pl.jit
    def k(
        x: pl.Tensor[[rows, COLS], pl.FP32],
        w: pl.Tensor[[1, HALF], pl.FP32],
        out: pl.Out[pl.Tensor[[rows, COLS], pl.FP32]],
    ):
        with pl.at(level=pl.Level.CORE_GROUP):
            t = x[0:rows, :]                    # [rows, 128] tile
            lo = t[:, 0:HALF]                   # column slice  <-- the defect
            hi = t[:, HALF:COLS]                # column slice  <-- the defect
            a = pl.col_expand_mul(lo, w)        # lo * w
            b = pl.col_expand_mul(hi, w)        # hi * w
            out = pl.assemble(out, pl.concat(a, b), [0, 0])
        return out
    return k


print(f"{'ROWS':>5}  {'max_abs_err':>12}  {'wrong elems':>14}  {'wrong rows':>12}  verdict")
print("-" * 68)
fails = []
for rows in (1, 2, 4, 8, 16):
    torch.manual_seed(0)
    x = torch.randn(rows, COLS, dtype=torch.float32)
    w = torch.randn(1, HALF, dtype=torch.float32)
    out = torch.zeros(rows, COLS, dtype=torch.float32)
    build(rows)(x, w, out, config=cfg)
    ref = x * torch.cat([w, w], dim=1)          # golden: every column scaled by w
    err = (out - ref).abs()
    bad = int((err > 1e-5).sum())
    badr = int((err.max(dim=1).values > 1e-5).sum())
    ok = bad == 0
    if not ok:
        fails.append(rows)
    print(f"{rows:>5}  {float(err.max()):>12.6f}  {bad:>6}/{rows*COLS:<7}  {badr:>5}/{rows:<6}  "
          f"{'PASS' if ok else '*** WRONG ***'}")

print()
if fails:
    print(f"REPRODUCED: column-slicing a multi-row tile is miscompiled for ROWS = {fails}.")
    print("ROWS = 1 passes (source and destination addresses coincide), which is exactly why")
    print("a single-row RoPE is unaffected while a 16-row one is destroyed.")
    # Show the signature on the largest failing case.
    rows = fails[-1]
    torch.manual_seed(0)
    x = torch.randn(rows, COLS, dtype=torch.float32)
    w = torch.randn(1, HALF, dtype=torch.float32)
    out = torch.zeros(rows, COLS, dtype=torch.float32)
    build(rows)(x, w, out, config=cfg)
    print(f"\nSignature (ROWS={rows}): out[0, 64:68] should be x[0, 64:68]*w[0,0:4]; the dense")
    print("repack instead leaves it holding row 1's LOWER half:")
    print(f"  device out[0, 64:68]      = {[round(v, 4) for v in out[0, HALF:HALF+4].tolist()]}")
    print(f"  correct  x[0,64:68]*w[:4] = "
          f"{[round(v, 4) for v in (x[0, HALF:HALF+4] * w[0, 0:4]).tolist()]}")
    print(f"  x[1, 0:4]*w[0:4]          = "
          f"{[round(v, 4) for v in (x[1, 0:4] * w[0, 0:4]).tolist()]}   <-- what actually lands")
    raise SystemExit(1)
print("NOT reproduced.")
```

</details>

## Blast radius

`t[:, a:b]` on a multi-row tile is a routine DSL construct. Any kernel doing it is silently wrong,
with no signal — and it is used identically by the *working* K path in the very same kernel, which is
what made this so hard to see.

Measured consequence in qwen3-14b's fused decode, grading against the un-fused kernel (a different
kernel, and the one that passes the serving e2e) on a fixture where every batch row shares one
`seq_len`:

| | `all_q_padded` vs a host oracle | bad rows | err@6 | argmax@6 |
|---|---|---|---|---|
| as-is | **5.765625** | **640/640** | 1.42 | 13/16 |
| with the Q RoPE forced to slice one row at a time | **0.000977** (a bf16 quantum) | **0/640** | 0.084 | **16/16** |

qwen3-14b's **serving e2e fails**. pypto-lib previously REVERTED the whole fused-decode feature
(pypto-lib#703, "Revert Qwen3 decode changes causing incorrect serving output") — this is one of the
reasons why.

## Suggested fix

The bad `MemRef` is already present at `29_after_InitMemRef`, i.e. **before** `MemoryReuse` and
`AllocateMemoryAddr` — so this is not a memory-reuse aliasing bug like #2007; the slice's shape is
mis-encoded at construction. Options:

1. **Give the sliced tile the parent's row pitch** (`pad = cols_parent - cols_slice`), making it a
   true strided view. `textract` then becomes a no-op and nothing is repacked. This looks like the
   right fix.
2. Or allocate a **genuinely disjoint** destination, so the strided -> dense repack is out-of-place.
3. Failing either, **reject it** — the precedent is `pto.alloc_multi_tile`, which refuses slot
   layouts that "would silently overlap". Silent wrong data is the worst possible outcome.

## Ruled out

Each was tested and falsified:

- **Not #2003 / #2007 / #2006** — reproduces with all three fixes applied.
- **Not `MemoryReuse`** — the mis-encoded MemRef exists before that pass runs.
- **Not UB over-subscription** — `fa_fused`'s report says Vec 98.5 KB / 184 KB = 53.5 %.
- **Not float reassociation, not the reduction tile shape, not the publish mechanism, not the
  fa_fused task structure** — each removed independently in the real kernel; none moved the error.
- **Not the K path** — proven bug-neutral by a Q/K bisect (unfolding K alone changes nothing;
  unfolding Q alone fixes it).

**Comment by luohuan19** [comment]
## Root cause, refined to the exact line

`src/ir/transforms/init_memref.cpp`:

```cpp
 56: bool IsViewOperation(const std::string& op_name) {
 59:   return registry.GetEntry(op_name).OutputMemoryInheritsInput();
 67: bool IsDataPermutingInheritOp(const OpPtr& op) { return IsOp(op, "tile.transpose"); }
...
361:     if (IsViewOperation(call->op_->name_) && call->args_.size() > 0) {
366:       const bool may_inherit = !IsDataPermutingInheritOp(call->op_);
```

`InitMemRef` treats `tile.slice` as a **pure metadata view** (`OutputMemoryInheritsInput()`), and the **only carve-out is `tile.transpose`** (line 67). So the slice result inherits the PARENT's buffer, at the byte offset of the slice origin computed with the **parent's** row pitch.

But the result tile is **dense with a different row pitch**, and the backend lowers the slice to `pto.textract` / `TEXTRACT` — a **physical strided → dense repack**. The repack therefore executes **over its own still-live source**.

Concretely, for `t = [16,128] FP32`, `lo = t[:, 0:64]`, `hi = t[:, 64:128]` (UB byte addresses from the emitted code): `t@256` (8192 B, pitch 512 B), `lo@256` (4096 B, dense pitch 256 B), `hi@512` (4096 B).

- `TEXTRACT(lo, t, 0, 0)` writes dest row *r* at `256 + 256r` while reading src row *r* at `256 + 512r`. Its **row-1 write** (bytes 512..768) **clobbers `t`'s row-0 upper half.**
- `TEXTRACT(hi, t, 0, 64)` then reads the already-trashed `t`: `hi` row 0 = `t` bytes [512, 768) = `x[1, 0:64]`.
- Device output matches exactly: `out[0, 64:68] == x[1, 0:4] * w[0:4]`.

**Row count is the whole story.** At `ROWS == 1`, dest addr == src addr, so the extract degenerates to an identity copy.

## Ablation — necessary and sufficient ingredients

Standalone, pypto-only (`repro_colslice2.py`; no pypto-lib, no `pl.spmd`, no `pl.split_aiv`, no `pl.system.syncall`, no `pl.pipeline`, no `pl.create_tensor`, no RoPE — ~20 lines of kernel per variant). Verbatim device output:

```
  variant                                       ROWS  max_abs_err  verdict
  ------------------------------------------------------------------------
  V1 two colslices of [R,128]  (== kernel RoPE)     1     0.000000  PASS
  V1 two colslices of [R,128]  (== kernel RoPE)    16     9.420689  *** WRONG ***
  V2 one colslice, parent unused after              1     0.000000  PASS
  V2 one colslice, parent unused after             16     0.000000  PASS
  V3 one colslice, parent READ after                1     0.000000  PASS
  V3 one colslice, parent READ after               16     5.503764  *** WRONG ***
  V4 row slice only (full col extent)              16     0.000000  PASS
  V5 row-slice-to-1-row FIRST, then colslice       16     0.000000  PASS
```

Necessary **and** sufficient:

1. a **tile-level** column slice, i.e. narrowing the **trailing** dim;
2. the parent tile has **rows > 1**;
3. the parent (or a second column slice of it) is **read AFTER** the slice.

**V2 is the key control**: the slice's *own* value comes out correct — what is destroyed is the **source** tile. V4 shows a *row* slice (pitch preserved) is safe. V5 is the kernel-side workaround.

## The fix

`tile.slice` must **not** inherit the parent's buffer when it narrows the trailing dim of a tile with `rows > 1` — i.e. whenever the dense result's row pitch differs from the parent's. It needs the same treatment `tile.transpose` already has at line 67, for the same reason, and the same `.not_inplace_safe()` reasoning that fixed #2007 (`tile.concat`).

Alternatively, give the sliced tile the **parent's** row pitch (`pad = cols_parent - cols_slice`), making it a genuine strided view — then `TEXTRACT` is a no-op and nothing is repacked at all. That is probably the better fix, since it also removes the copy.
